docs/how-to-write-a-module.md

```diff
@@ -0,0 +1,257 @@
+
+We recommend using a file per a logical area. This approach makes it easier to maintain the module.
+
+Suggested minimal layout:
+
+| Filename                                          | Contains                                               |
+| ------------------------------------------------- |------------------------------------------------------- |
+| [`module_name.go`](#file-module_namego)           | Module configuration, implementation and registration. |
+| [`charts.go`](#file-chartsgo)                     | Charts, charts templates and constructor functions.    |
+| [`init.go`](#file-initgo)                         | Initialization methods.                                |
+| [`collect.go`](#file-collectgo)                   | Metrics collection implementation.                     |
+| [`module_name_test.go`](#file-module_name_testgo) | Public methods/functions tests.                        |
+| [`testdata/`](#file-module_name_testgo)           | Tests fixtures.                                        |
+
+### File `module_name.go`
+
+> :exclamation: See the example [`examlpe.go`](https://github.com/netdata/go.d.plugin/blob/master/modules/example/example.go).
+
+Don't overload this file with the implementation details.
+
+Usually it contains only:
+ 
+-   module registration.
+-   module configuration.
+-   [module interface implementation](#module-interface).
+
+### File `charts.go`
+
+> :exclamation: See the example: [`charts.go`](https://github.com/netdata/go.d.plugin/blob/master/modules/example/charts.go).
+
+Put charts, charts templates and charts constructor functions in this file. 
+
+### File `init.go`
+
+> :exclamation: See the example: [`init.go`](https://github.com/netdata/go.d.plugin/blob/master/modules/example/init.go).
+
+All the module initialization details should go in this file.
+
+-   make a function for each value that needs to be initialized.
+-   a function should return a value(s), not implicitly set/change any values in the main struct.
+
+```go
+// init.go
+
+// Prefer this approach.
+func (e Example) initSomeValue() (someValue, error) {
+    // ...
+    return someValue, nil 
+}
+
+// This approach is ok too, but we recommend to not use it.
+func (e *Example) initSomeValue() error {
+    // ...
+    m.someValue = someValue
+    return nil
+}
+```     
+
+### File `collect.go`
+
+> :exclamation: See the example: [`collect.go`](https://github.com/netdata/go.d.plugin/blob/master/modules/example/collect.go).
+
+This file is the entry point for the metrics collection.
+
+Feel free to split it into several files if you think it makes the code more readable.
+
+Use `collect_` prefix for the filenames: `collect_this.go`, `collect_that.go`, etc.
+
+```go
+// collect.go
+
+func (e *Example) collect() (map[string]int64, error) {
+    collected := make(map[string])int64
+    // ...
+    // ...
+    // ...
+    return collected, nil
+}
+```
+
+### File `module_name_test.go`
```

Comment:
Can we add some links to resources for creating tests in go?  Can we give some examples from already implemented collectors? 
Examples would be test code + explanation/description of what it does. I want even the go rookie programmer who hasn't written any tests to quickly understand the concept/logic and create the tests without giving up on the collector.
Remember, we want to nudge people from creating a quick and dirty python collector to using go which is more complex but more robust as well. 
Perhaps @kaskavel can help with this one?

Reply:
@OdysLam I will take a look now on this. Let me go through the existing documentation and I 'll post here.